File: Graph/1707번 - 이분 그래프/1707.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

#define LIM 20001
#define RED 2
#define BLUE 3

int tc;
int v, e;
vector<int> edges[LIM];

int color[LIM] = { 0, };

void all_clear() {
    fill_n(&color[0], LIM, 0);
    for(int i = 1; i < LIM; i++) {
        edges[i].clear();
    }
}

void input() {
    cin >> v >> e;
    
    all_clear();
    
    for(int i = 0, a, b; i < e; i++) {
        cin >> a >> b;
        edges[a].push_back(b);
        edges[b].push_back(a);
    }
}

bool bfs(int st) {
    queue<int> Q;
    Q.push(st);
    color[st] = RED;
    
    while(!Q.empty()) {
        int cur = Q.front(); Q.pop();
        int nxt_color = (color[cur] == RED ? BLUE : RED);
        
        for(int i = 0; i < edges[cur].size(); i++) {
            int nxt = edges[cur][i];
            if(!color[nxt]) {
                color[nxt] = nxt_color;
                Q.push(nxt);
            } else if(color[cur] == color[nxt]) {
                return false;
            }
        }
    }
    
    return true;
}

void solve() {
    bool ans = true;
    for(int i = 1; i <= v; i++) {
        if(!color[i]) ans = bfs(i);
        // 만일, 이분그래프가 아닐 경우, 바로 종료
        if(!ans) break;
    }
    
    cout << (ans ? "YES" : "NO") << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    cin >> tc;
    while(tc--) {
        input();
        solve();
    }
    
    return 0;
}

File: Graph/1707번 - 이분 그래프/1707.java
import java.io.*;
import java.util.*;

public class Main {
    static final int LIM = 20001;
    static final int RED = 2;
    static final int BLUE = 3;

    static int tc;
    static int v, e;
    static List<Integer>[] edges = new ArrayList[LIM];
    static int[] color = new int[LIM];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        tc = Integer.parseInt(br.readLine());

        for (int i = 0; i < LIM; i++) {
            edges[i] = new ArrayList<>();
        }

        while (tc-- > 0) {
            input(br);
            solve();
        }
    }

    static void input(BufferedReader br) throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        v = Integer.parseInt(st.nextToken());
        e = Integer.parseInt(st.nextToken());
        
        // 이전 입력값 초기화
        Arrays.fill(color, 0);
        for (int i = 1; i <= v; i++) {
            edges[i].clear();
        }

        for (int i = 0; i < e; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            edges[a].add(b);
            edges[b].add(a);
        }
    }
    
    static void solve() {
        boolean isBipartite = true;

        for (int i = 1; i <= v; i++) {
            if (color[i] == 0) {
                if (!bfs(i)) {
                    isBipartite = false;
                    break;
                }
            }
        }

        System.out.println(isBipartite ? "YES" : "NO");
    }

    static boolean bfs(int start) {
        Queue<Integer> q = new LinkedList<>();
        q.add(start);
        color[start] = RED;

        while (!q.isEmpty()) {
            int cur = q.poll();
            int nextColor = (color[cur] == RED) ? BLUE : RED;

            for (int next : edges[cur]) {
                if (color[next] == 0) {
                    color[next] = nextColor;
                    q.add(next);
                } else if (color[next] == color[cur]) {
                    return false;
                }
            }
        }
        return true;
    }
}
